pages/plugins/writing.md.erb

```diff
@@ -0,0 +1,225 @@
+        name: '*.md'
+```
+
+{:toc}
+
+## Create a new git repository
+
+Every Buildkite plugin is a Git repository, ending in `-buildkite-plugin`. Let's create a new Git repository following these naming conventions:
+
+```shell
+mkdir file-counter-buildkite-plugin
+cd file-counter-buildkite-plugin
+git init
+```
+
+## Add a plugin.yml
+
+Next we'll create the first file: a `plugin.yml` file. This describes how it will appear in the [Buildkite Plugin Directory](/docs/plugins/directory).
+
+```shell
+touch plugin.yml
+```
+
+```yaml
+name: File Counter
+description: Annotates the build with a file count
+author: https://github.com/a-github-user
+requirements: []
+configuration:
+  pattern:
+    name: string
+  additionalProperties: false
+```
+
+The `configuration` property defines the validation rules for the plugin configuration using the [JSON Schema](https://json-schema.org) format. We’ve specified that the plugin has a single `name` property, of type `string`.
+
+## Add a hook
+
+Plugins can implement a number of [plugin hooks](/docs/agent/v3/hooks). For this plugin, we’ll create a `post-command` hook in a `hooks` directory:
+
+```shell
+mkdir hooks
+touch hooks/post-command
+chmod +x hooks/post-command
+```
+
+```shell
+#!/bin/bash
+set -euo pipefail
+
+PATTERN="$BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN"
+
+echo "--- \:1234\: Counting the number of files"
+
+COUNT=$(find . -name "$PATTERN" | wc -l)
+
+echo "Found ${COUNT} files matching ${PATTERN}"
+
+buildkite-agent annotate "Found ${COUNT} files matching ${PATTERN}"
+```
+
+## Add a test
+
+Next step is to test the `post-command` hook using BATS, and the the `buildkite/plugin-tester` Docker image.
+
+```shell
+mkdir tests
+touch tests/post-command.bats
+chmod +x tests/post-command.bats
+```
+
+Create the following `tests/post-command.bats` file:
+
+```shell
+#!/usr/bin/env bats
+
+load '/usr/local/lib/bats/load.bash'
+
+@test "Creates an annotation with the file count" {
+  export BUILDKITE_PLUGIN_FILE_COUNTER_PATTERN="*.bats"
+
+  stub buildkite-agent annotate "Found 1 files matching *.bats"
+
+  run "$PWD/hooks/post-command"
+
+  unstub buildkite-agent
+
+  assert_success
+  assert_output --partial "Found 1 files matching *.bats"
+}
+```
+
+To run the test, run the following Docker command:
+
+```shell
+docker run -it --rm -v "$(PWD):/plugin:ro" buildkite/plugin-tester
+```
+
+```
+ ✓ Creates an annotation with the file count
+
+1 test, 0 failures
+```
+
+To make it easier to run this command, create a Docker Compose file:
+
+```shell
+touch docker-compose.yml
+```
+
+```yml
+version: '2'
+services:
+  tests:
+    image: buildkite/plugin-tester
+    volumes:
+      - ".:/plugin:ro"
+```
+
+You can now run the tests using the following command:
+
+```shell
+docker-compose run --rm tests
+```
+
+## Add a readme
+
+Next we'll add a `README.md` file to introduce the plugin to the world:
+
+```shell
+touch README.md
+```
+
+<%= render 'plugins/tutorial_readme' %>
```

Comment:
Does this render markdown? or html? I wonder if we can easily code highlight it.

Reply:
I wanted it to be syntax highlighted too! Extracting that out, into a separate file, is actually a hack itself… because nesting markdown in markdown breaks the renderer. The real fix was switching to Commonmarker instead of RedCarpet but that was harder than I hoped.

I reckon it's a nice to have, but we should take another stab at seeing if we can make it syntax highlighted!